include/eeprom/simulation/eeprom.hpp

```diff
@@ -67,10 +79,164 @@ class EEProm : public I2CDeviceBase,
+                    BACKING_SIZE, BACKING_SIZE - start, start);
+            }
+            // pointer back to 0, though we'll be doing a lot of SEEK_SET so it
+            // doesn't really matter
+            std::fseek(file, 0, SEEK_SET);
+            LOG("got file at %p\n", file);
+            return file;
+        }
+        static auto get_prepped_backing_file(const po::variables_map& variables)
+            -> FILE* {
+            return get_prepped_backing_file(
+                variables["eeprom-filename"].as<std::string>());
+        }
+        static auto add_options(po::options_description& cmdline,
+                                po::options_description& env)
+            -> std::function<std::string(std::string)> {
+            cmdline.add_options()(
+                "eeprom-filename,f",
+                po::value<std::string>()->default_value(
+                    std::string(TEMPFILE_KEY)),
+                "path to backing file for eeprom. if unspecified, a temp will "
+                "be used. May be specified in an environment file called "
+                "EEPROM_FILENAME.");
+            env.add_options()("EEPROM_FILENAME",
+                              po::value<std::string>()->default_value(
+                                  std::string(TEMPFILE_KEY)));
+            return [](std::string input_val) -> std::string {
+                if (input_val == "EEPROM_FILENAME") {
+                    return "eeprom-filename";
+                }
+                return "";
+            };
+        }
+        auto read(uint8_t* readbuf, uint16_t address, size_t size) -> void {
+            if (std::fseek(backing, address, SEEK_SET)) {
+                fprintf(stderr,
+                        "fseek to %d for EEPROM read on file %p failed: eof=%d "
+                        "err=%d %d %s",
+                        address, backing, std::feof(backing),
+                        std::ferror(backing), errno, strerror(errno));
+            }
+            auto read_count =
+                std::fread(readbuf, sizeof(readbuf[0]), size, backing);
+            if (read_count != size) {
+                if (std::feof(backing)) {
+                    fprintf(stderr,
+                            "EEPROM read of %ld bytes from %d read only %ld "
+                            "and hit eof\n",
+                            size, address, read_count);
+                    std::abort();
+                } else if (std::ferror(backing)) {
+                    fprintf(stderr,
+                            "EEPROM read of %ld bytes from %d on file %p read "
+                            "only %ld hit error: %d %s\n",
+                            size, address, backing, read_count, errno,
+                            strerror(errno));
+                    std::abort();
+                }
+            }
+        }
+
+        auto write(const uint8_t* writebuf, uint16_t address, size_t size)
```

Comment:
Small thing, right now we don't attempt to catch when writing over a page boundry on the older chips which causes it to wrap to the same page, not the next page.
in include/eeprom/core/task.hpp we have:
if (hw_iface.get_eeprom_chip_type() ==
                hardware_iface::EEPromChipType::MICROCHIP_24AA02T &&
            ((m.memory_address % 8) + m.length) > 8) {
            LOG("Warning: write request will overrun page");
        }
        
If you want the simulator to be exact to that hardware when the hardware is the MICROCHIP_24AA02T (which is the current default) you'd have to catch and reproduce that behavior. Or we can edit task.hpp to actually handle these writes across pages. This shouldn't happen if you're using one of my accessor classes but if someone was trying to directly control the eeprom from Can messages or something it could happen

Reply:
Oh interesting. That definitely might happen here (though I'm not that worried about the older chips since we're gonna stop using them) but i think we'd want to put that in the main sim eeprom class, not the backing storage.